content/post/walkthrough-exploiting-bugs-to-gain-code-execution.md

```diff
@@ -0,0 +1,338 @@
+
+But as an exercise, I occasionally feel a need to actually develop exploits, and if you're into C++, it's a great learning experience. Due to advances in binary security there are an increasing number of hurdles, but I'm gonna walk you through the steps to exploit a simple overflow on a modern operating system.
+
+**Prerequisites:** Basic knowledge of how a [call stack](https://en.wikipedia.org/wiki/Call_stack) works, basic understanding of common assembly instructions (push, pop, mov, add...). Being comfortable debugging executables is a plus.
+
+My payload will be for OSX Yosemite, but the same techniques apply to other systems.
+
+The vulnerability
+-----------------
+
+First, we'll need a vulnerable program. I'm gonna write up a quick one for this, but if you want to go hunting for these kinds of things in the wild, by far the easiest way is to grab a copy of [AFL](http://lcamtuf.coredump.cx/afl/) and start fuzzing open source projects.
+
+Our vulnerable program: https://gist.github.com/ccbrown/ff9925a58c2e45da2b5b
+
+The important part is here:
+
+```
+class ConnectionHandler {
+public:
+    ConnectionHandler(int fd) : _fd(fd) {}
+    ~ConnectionHandler() { close(_fd); }
+
+    int run() {
+        while (true) {
+            ssize_t bytes = recv(_fd, _recvBuffer + _recvBytes, sizeof(_recvBuffer), 0);
+            if (bytes <= 0) {
+                break;
+            }
+            _recvBytes += bytes;
+            void* newline = memchr(_recvBuffer, '\n', _recvBytes);
+            if (newline) {
+                send(_fd, _recvBuffer, (char*)newline - _recvBuffer + 1, 0);
+                break;
+            }
+        }
+        return 0;
+    }
+
+private:
+    int _fd = 0;
+    char _recvBuffer[512];
+    size_t _recvBytes = 0;
+};
+```
+
+See the bug? Good! Does it look like the kind of mistake you've made before? Maybe, maybe not... but let's pwn it. 
+
+The plan of attack
+------------------
+
+This server echos the first line you send to it, then closes the connection. But there's a run-of-the-mill buffer overflow.
+
+So what can we overwrite with our data? The thing of primary interest here is `_recvBytes`. We can manipulate that to overwrite data at any address we choose relative to `_recvBuffer`.
+
+Traditionally, we would stick our shell code into `_recvBuffer`, then write over a return address on the stack in order to execute it. But any modern system will make the stack and heap both non-executable. The trick here is to use [return-oriented programming](https://en.wikipedia.org/wiki/Return-oriented_programming).
+
+Writing the shell code
+----------------------
+
+First, let's sketch out our shell code in C. The goal here is to be very minimal.
```

Comment:
Why minimal? I know why, but I think it might be worth adding a quick phrase to clarify. Maybe it's unnecessary?